### ISSUE TITLE
Export model in PyTorch format

### ISSUE THREAD

**Issue by philippart-s** [ultralytics/ultralytics#3396]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When I use the export function on the model without any argument, the function export in torchscript format insteaf PyTorch format as mentioned in the documentation (https://docs.ultralytics.com/modes/export/#export-formats):
```bash
pathToModel = model.export()

Ultralytics YOLOv8.0.111 🚀 Python-3.9.5 torch-2.0.1+cu117 CUDA:0 (Tesla V100S-PCIE-32GB, 32510MiB)
Model summary (fused): 168 layers, 3006233 parameters, 0 gradients

PyTorch: starting from runs/detect/train4/weights/best.pt with input shape (16, 3, 640, 640) BCHW and output shape(s) (16, 7, 8400) (6.0 MB)

TorchScript: starting export with torch 2.0.1+cu117...
TorchScript: export success ✅ 1.3s, saved as runs/detect/train4/weights/best.torchscript (11.9 MB)

Export complete (1.5s)
Results saved to /workspace/101-ai-lab-sources/src/notebook/runs/detect/train4/weights
Predict:         yolo predict task=detect model=runs/detect/train4/weights/best.torchscript imgsz=640 
Validate:        yolo val task=detect model=runs/detect/train4/weights/best.torchscript imgsz=640 data=/workspace/data/rock-paper-scissors/data.yaml 
Visualize:       https://netron.app/
```

Have you any ideas what I'm doing wrong?

Thanks :)

**Comment by philippart-s** [comment]
> @philippart-s pytorch is the native model format. You can export to a pytorch model to any other format using `yolo export`.
> 
> If you just need a pytorch model you've already got it.

Thanks @glenn-jocher for your answer.
What I want to do is to save the best model (best.pt) after training in a defined path.
As during the training the path is dynamic, `runs/detect/train4/weights/best.pt` in my example I want to save the model to another path after my training and, for this, I used `export` function but perhaps I used the wrong function?

**Comment by glenn-jocher** [comment]
@philippart-s Thanks for the follow-up question. The `export` function in YOLOv8 is used to convert the model to a production-ready format, which could be used for inference or deployment. What you are looking for is to save the best model after training. You can save the best model using the `save` function in YOLOv8.

Here is an example code snippet:
```
# train the model
model.train()

# save the best model after training
best_model_path = "/path/to/save/best/model.pt"
model.save(best_model_path)
```

This will save the best model to the specified path after the training. Please note that the `model.save` function will only save the weights and the model architecture and not the optimizer or the training state. Thus, to resume the training from the saved model, you will need to load the weights of the saved model to the corresponding layers of the model and initialize the optimizer and scheduler to their respective states.

**Comment by philippart-s** [comment]
Thanks for your explanations but when I try to save my model I have an error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 model.save("/workspace/model/rock-paper-scissors/model2.pt")

File ~/.miniconda3/lib/python3.9/site-packages/ultralytics/yolo/engine/model.py:116, in YOLO.__getattr__(self, attr)
    114 """Raises error if object has no requested attribute."""
    115 name = self.__class__.__name__
--> 116 raise AttributeError(f"'{name}' object has no attribute '{attr}'. See valid attributes below.\n{self.__doc__}")

AttributeError: 'YOLO' object has no attribute 'save'. See valid attributes below.

    YOLO (You Only Look Once) object detection model.

    Args:
        model (str, Path): Path to the model file to load or create.
        task (Any, optional): Task type for the YOLO model. Defaults to None.

    Attributes:
        predictor (Any): The predictor object.
        model (Any): The model object.
        trainer (Any): The trainer object.
        task (str): The type of model task.
        ckpt (Any): The checkpoint object if the model loaded from *.pt file.
        cfg (str): The model configuration if loaded from *.yaml file.
        ckpt_path (str): The checkpoint file path.
        overrides (dict): Overrides for the trainer object.
        metrics (Any): The data for metrics.

    Methods:
        __call__(source=None, stream=False, **kwargs):
            Alias for the predict method.
        _new(cfg:str, verbose:bool=True) -> None:
            Initializes a new model and infers the task type from the model definitions.
        _load(weights:str, task:str='') -> None:
            Initializes a new model and infers the task type from the model head.
        _check_is_pytorch_model() -> None:
            Raises TypeError if the model is not a PyTorch model.
        reset() -> None:
            Resets the model modules.
        info(verbose:bool=False) -> None:
            Logs the model info.
        fuse() -> None:
            Fuses the model for faster inference.
        predict(source=None, stream=False, **kwargs) -> List[ultralytics.yolo.engine.results.Results]:
            Performs prediction using the YOLO model.

    Returns:
        list(ultralytics.yolo.engine.results.Results): The prediction results.
```

**Comment by glenn-jocher** [comment]
@philippart-s thanks for reaching out! It seems that the `save` method is not available in the `YOLO` object of YOLOv8. The error message indicates that the attribute `'save'` is not recognized.

To save the model in YOLOv8, you can instead use the `export` function. This function allows you to export the model to a production-ready format that can be used for inference or deployment.

To export the model, you can use the following code snippet:
```python
model.export('/path/to/save/model.pt')
```

Replace `'/path/to/save/model.pt'` with the desired path and filename where you want to save the model.

Please note that the `export` function will save the model in a format that is compatible with PyTorch and can be loaded using the `torch.load` function.

**Comment by philippart-s** [comment]
I'm not sure that the `export` function allow to set a path, giving the documentation (https://docs.ultralytics.com/modes/export/#arguments) the `export` function have not parameter to set the path and when I use it as you mentioned I have the following error:

```
model.export('/path/to/save/model.pt')

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[20], line 1
----> 1 model.export('/path/to/save/model.pt')

TypeError: export() takes 1 positional argument but 2 were given
```

If I understand the documentation I can export the model, for example in TF format,  with this code:
```python
exportedModelMetaData = model.export(format='saved_model')
```
But this code store the exported model in an arbitrary path.
What I want since the beginning is to choose the path where my model is saved / exported in Python to automatise the training and after the ready to prod deployment model.

**Comment by philippart-s** [comment]
An update after research and tests, I think perhaps the `export` function has a bug or a non deterministic behavior.
If I understand the code of the function the return value must be the path to the exported model, it's the case for `onnx` and `torchscript` for example but not for `save_model` (TF), is it a bug?

The tests:
```bash
# Export the model
exportedModelMetaData = model.export(format='onnx')
print(exportedModelMetaData)

-> runs/detect/train3/weights/best.onnx
```

```bash
# Export the model
exportedModelMetaData = model.export(format='torchscript')
print(exportedModelMetaData)

-> runs/detect/train3/weights/best.torchscript
```

```bash
# Export the model
exportedModelMetaData = model.export(format='saved_model')
print(exportedModelMetaData)

-> []
```

**Comment by philippart-s** [comment]
Ok I find the pb, I used an "old" version 8.0.111, after upgrade to version 8.0.124 the function return the path with the parameter `save_model`.

Then, I think I've just to copy the exported model to the desired path in my Python development.

**Comment by glenn-jocher** [comment]
@philippart-s i'm glad you were able to resolve the issue by updating to the latest version of YOLOv8. After exporting the model using the `export` function with the `save_model` format, you can indeed copy the exported model to the desired path in your Python development using standard file operations or libraries like `shutil` in Python.

To automate the process, you can use the `shutil.copy` function to copy the exported model from the original path to the desired path in your Python script. Here's an example:

```python
import shutil

exported_model_path = "runs/detect/train3/weights/best.tf"  # Path of the exported model
desired_path = "/path/to/desired/model/model.tf"  # Desired path to save the model

shutil.copy(exported_model_path, desired_path)
```

Replace `"runs/detect/train3/weights/best.tf"` with the actual path of the exported model, and `"/path/to/desired/model/model.tf"` with the desired path and filename where you want to save the model.

By using the `shutil.copy` function, you can copy the exported model to the desired path during your Python script execution, automating the process of saving the model to a specific location.

**Comment by philippart-s** [comment]
Thanks, I close the issue

**Comment by glenn-jocher** [comment]
@philippart-s hello,
Thank you for reaching out to us. We apologize for any confusion or inconvenience caused. 

In YOLOv8, the `export` function allows you to export the model to different formats like ONNX, TorchScript, and SavedModel. However, there was a bug in the previous version that caused the `save_model` format to not return the exported model's path.

To resolve this issue, we released an updated version (8.0.124) of YOLOv8, which fixed the bug. After upgrading to this version, the `export` function with the `save_model` format should correctly return the path to the exported model.

To save the exported model to a specific location, you can use standard file operations or libraries like `shutil` in Python to copy the exported model from its original path to the desired path. This can be done programmatically in your Python script.

We appreciate your understanding and apologize for any inconvenience caused by the earlier version. Please let us know if you have any further questions or concerns.

**Comment by glenn-jocher** [comment]
@pfeatherstone in YOLOv8, the export process typically involves fusing layers to optimize the model for deployment. If you need to export the model without fusing, you can refer to the export documentation and use the provided options to customize your export process.
